components/board/genericlinux/digital_interrupts.go

```diff
@@ -6,115 +6,159 @@ package genericlinux
+}
 
-	cancelCtx, cancelFunc := context.WithCancel(ctx)
-	result := digitalInterrupt{
-		boardWorkers: &b.activeBackgroundWorkers,
-		interrupt:    interrupt,
-		line:         line,
-		cancelCtx:    cancelCtx,
-		cancelFunc:   cancelFunc,
-		config:       &config,
+func (di *digitalInterrupt) Close() error {
+	di.workers.Stop()
+
+	// The background worker has now stopped, so cannot have locked the mutex. Anything else cannot
+	// lock the mutex for very long, so we'll be able to acquire it quickly here.
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	var err error
+	if len(di.channels) > 0 {
+		err = fmt.Errorf("closed digital interrupt %s, but it still had %d listeners",
+		                 di.config.Name, len(di.channels))
 	}
-	result.startMonitor()
-	return &result, nil
+	return multierr.Combine(err, di.line.Close())
+}
+
+func (di *digitalInterrupt) Name() string {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	return di.config.Name
 }
 
-func (di *digitalInterrupt) startMonitor() {
-	di.boardWorkers.Add(1)
-	utils.ManagedGo(func() {
-		for {
-			select {
-			case <-di.cancelCtx.Done():
-				return
-			case event := <-di.line.Events():
-				utils.UncheckedError(pinwrappers.Tick(
-					di.cancelCtx, di.interrupt.(*pinwrappers.BasicDigitalInterrupt), event.RisingEdge, uint64(event.Time.UnixNano())))
+func (di *digitalInterrupt) Value(
+	ctx context.Context,
+	extra map[string]interface{},
+) (int64, error) {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	return di.count, nil
+}
+
+func (di *digitalInterrupt) monitor(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event := <-di.line.Events():
+			di.mu.Lock()
+			defer di.mu.Unlock()
+
+			if event.RisingEdge {
+				di.count++
+			}
+
+			tick := board.Tick{
+				Name: di.config.Name,
+				High: event.RisingEdge,
+				TimestampNanosec: uint64(event.Time.UnixNano()),
+			}
+			for _, ch := range di.channels {
+				select {
+				case <-ctx.Done():
+					return
+				case ch<-tick:
+				}
 			}
 		}
-	}, di.boardWorkers.Done)
+	}
 }
 
-func closeInterrupt(di *digitalInterrupt) error {
-	// We shut down the background goroutine that monitors this interrupt, but don't need to wait
-	// for it to finish shutting down because it doesn't use anything in the line itself (just a
-	// channel of events that the line generates). It will shut down sometime soon, and if that's
-	// after the line is closed, that's fine.
-	di.cancelFunc()
-	return di.line.Close()
+// AddChannel adds a board.Tick channel to stream ticks to.
+func (di *digitalInterrupt) AddChannel(ch chan board.Tick) {
+	di.mu.Lock()
+	defer di.mu.Unlock()
+	di.channels = append(di.channels, ch)
 }
 
-// struct implements board.GPIOPin to support reading current state of digital interrupt pins as GPIO inputs.
-type gpioInterruptWrapperPin struct {
-	interrupt digitalInterrupt
+// RemoveChannel removes a previously-added channel to stream ticks to.
+func (di *digitalInterrupt) RemoveChannel(ch chan board.Tick) {
```

Comment:
why is this called on the interrupt instead of being a helper? will this work with modules?

Reply:
How would you implement `b.StreamTicks()` without this? We need some way to do something similar to [this old line](https://github.com/viamrobotics/rdk/pull/3995/files#diff-a6ace4197b4a5a2f76d35d720a859aae92059216ed4bc7371f0d211f9dbfdd38L470) that was replaced by [this new line](https://github.com/viamrobotics/rdk/pull/3995/files#diff-a6ace4197b4a5a2f76d35d720a859aae92059216ed4bc7371f0d211f9dbfdd38R482). 

> will this work with modules?

Yes, by the time you get into `b.StreamTicks()`, we're already agnostic to whether the receiving side of the channel is sending things to a different module or not.

Reply:
I agree that we need but, im referring to the way that the old line has the interrupt as an argument while this new one is being called on the client. I know that I ran into issues with modules when the interrupt was a `digitalinterruptClient` but maybe its not an issue here.

Reply:
Hm... I should double-check this! When we're both in the office, we should talk about how to set it up with modules to see whether things continue to work.

Reply:
yeah sounds good, we can test with the local modules I have tomorrow